source/Registry.cpp

```diff
@@ -298,14 +347,18 @@ void Registry::dump_models(
         // Write the current batch of models to file.
         auto writer = JsonValidation::compact_writer();
         for (std::size_t i = batch_size * batch; i < batch_size * (batch + 1) &&
-             i < models.size() + field_models.size();
+             i < models.size() + field_models.size() + literal_models.size();
              i++) {
           if (i < models.size()) {
             writer->write(models[i].to_json(context_), &batch_stream);
-          } else {
+          } else if (i < models.size() + field_models.size()) {
             writer->write(
                 field_models[i - models.size()].to_json(context_),
                 &batch_stream);
+          } else {
+            writer->write(
+                literal_models[i - models.size() - field_models.size()].to_json(context_),
+                &batch_stream);
```

Comment:
pre-existing but we should probably rewrite this with a proper iterator. These index computations are error prone.

Reply:
Just to be clear, I'm not expecting you to implement this. This could be done in another PR, or we could do this internally.

Anyway, I changed my mind here. Yes we could probably make this work with iterators but the fact that iterators hide the indexing will make this code unreadable.

What might make it easier to read is abstracting away "models + field_models + literal_models". We could have:
```
class ModelVector {
  public:
   std::size_t size() const { return _models.size() + _field_models.size() + _literal_models.size(); }

  Json::Value json_for_index(std::size_t i) {
    if (i < _models.size()) {
      return _models[i].to_json();
    }
    i -= _models.size();
    if (i < _field_models.size()) {
      return _field_models[i].to_json();
    }
    i -= _field_models.size();
    if (i < _literal_models.size()) {
      return _literal_models[i].to_json();
    }
    mt_unreachable("invalid index");
  }

  private:
    std::vector<Model> _models;
    std::vector<FieldModel> _field_models;
    std::vector<LiteralModel> _literal_models;
}
```